From 129e006f821b318f4e82fc35c57d4b0e138a40c9 Mon Sep 17 00:00:00 2001
From: Jaeyoon Jung <jaeyoon.jung@lge.com>
Date: Tue, 2 Jan 2018 08:36:18 +0900
Subject: [PATCH] QQuickTextEdit: Fix cursor movement in compose mode

In the compose mode, the left key moves the cursor to the beginning of
the preedit text. For the right key, move the cursor by 1 step forward
from the end of the preedit text.

Change-Id: I1059f426a68ee1215e1047c0892d7f471a98f8af
Upstream-Status: Submitted [https://codereview.qt-project.org/215470]
---
 src/quick/items/qquicktextinput.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/quick/items/qquicktextinput.cpp b/src/quick/items/qquicktextinput.cpp
index aac7137ff3..caed0e8332 100644
--- a/src/quick/items/qquicktextinput.cpp
+++ b/src/quick/items/qquicktextinput.cpp
@@ -1511,6 +1511,11 @@ void QQuickTextInput::keyPressEvent(QKeyEvent* ev)
         if (!ignore && cursorPosition == d->m_text.length())
             ignore = ev->key() == (d->layoutDirection() == Qt::LeftToRight ? Qt::Key_Right : Qt::Key_Left);
     }
+#ifndef QT_NO_IM
+    // Process left/right in the compose mode.
+    if ((ev->key() == Qt::Key_Right || ev->key() == Qt::Key_Left) && d->composeMode())
+        ignore = false;
+#endif
     if (ignore) {
         ev->ignore();
     } else {
@@ -4477,6 +4482,10 @@ void QQuickTextInputPrivate::processKeyEvent(QKeyEvent* event)
     else if (event == QKeySequence::MoveToNextChar) {
         if (hasSelectedText()) {
             moveCursor(selectionEnd(), false);
+#ifndef QT_NO_IM
+        } else if (composeMode()) {
+            cursorForward(0, preeditAreaText().length() + 1);
+#endif
         } else {
             cursorForward(0, visual ? 1 : (layoutDirection() == Qt::LeftToRight ? 1 : -1));
         }
@@ -4487,6 +4496,10 @@ void QQuickTextInputPrivate::processKeyEvent(QKeyEvent* event)
     else if (event == QKeySequence::MoveToPreviousChar) {
         if (hasSelectedText()) {
             moveCursor(selectionStart(), false);
+#ifndef QT_NO_IM
+        } else if (composeMode()) {
+            cursorForward(0, 0);
+#endif
         } else {
             cursorForward(0, visual ? -1 : (layoutDirection() == Qt::LeftToRight ? -1 : 1));
         }
